test/Microsoft.AspNetCore.Authentication.Test/CollectingEventListener.cs

```diff
@@ -0,0 +1,74 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Authentication.Test
+{
+    // REVIEW: Also to go to some testing or shared-source assembly
+    // This can only collect from EventSources created AFTER this listener is created
+    public class CollectingEventListener : EventListener
+    {
+        private ConcurrentQueue<EventWrittenEventArgs> _events = new ConcurrentQueue<EventWrittenEventArgs>();
+        private HashSet<string> _eventSources;
+
+        public IReadOnlyList<EventWrittenEventArgs> EventsWritten => _events.ToArray();
+
+        public CollectingEventListener(params string[] eventSourceNames)
```

Comment:
So. Here's the thing with `EventListener`. Because `EventSource`s can be created at any time, `OnEventSourceCreated` is used to make `EventListener`s "work". At construction time, it will be called once for each existing `EventSource`, then it will be called as each new `EventSource` is initialized. However, this is all done by the base class constructor, which means that when `OnEventSourceCreated` is called for `EventSources` that **already exist** at the time of construction, the derived constructor body **has not yet been run**, meaning `_eventSources` is unavailable.

This "works" right now in that I only tested running the single test I added, which means that the `AuthenticationEventSource` was only created **after** the `EventListener` was initialized. However, if it ran as part of a batch of tests where any previous test initialized the `AuthenticationEventSource`, it would all fall apart.

I'm not a fan of the idea of creating a listener that hooks every `EventSource` as there is a noticible perf impact to enabling any single `EventSource` and there are a few in the framework (plus as we add them to our components...).

One option is to just make `CollectingEventListener` abstract and require creating subclasses that hard-code the sources they care about.

Another option would be to make the `AuthenticationEventSource` (and friends) "pubternal" (or even public with all `internal` methods). `OnEventSourceCreated` only needs to be used when your listener doesn't have access to the actual `EventSource` type, you only have the name. If they do have that type, they can just call `.EnableEvents` and pass it in. I'm kinda partial to this as it would also make building arbitrary consumers a little easier (when the consumer is willing to reference our code directly) as they have a concrete type to use instead of a stringly-typed name.

Reply:
We could also try to build something over `IObservable` like the SLAB of yore (https://msdn.microsoft.com/en-us/library/dn440729(v=pandp.60).aspx).

Reply:
@pakrym and I may have a solution using devious locks :)

Reply:
Solution found!